addons/l10n_it_edi/models/account_invoice.py

```diff
@@ -194,6 +195,85 @@ def format_alphanumeric(text_to_convert):
             if rc_refund:
                 document_total = -abs(document_total)
 
+        # List of dictionaries passed to the template for the part of 'DettaglioLinee'
+        invoice_lines = []
+        lines = self.invoice_line_ids.filtered(lambda l: not l.display_type)
+        for num, line in enumerate(lines):
+
+            # In the case of reverse_charge, the unit price and total price should be in euros, we assume the balance to be in Euros because
+            # only Italian companies can perform reverse charge, and they are legally required to use euros.
+            if reverse_charge:
+                price_subtotal = line.balance
+            else:
+                price_subtotal = line.price_subtotal
+
+            # In the case of reverse charge refund and downpayments, the values for the unit price and total price should be negative
+            if (line._get_downpayment_lines() and document_type != "TD02") or rc_refund:
+                price_subtotal = -abs(price_subtotal)
+            else:
+                price_subtotal = abs(price_subtotal)
+
+            # Unit price
+            unit_price = 0
+
+            if line.quantity and line.discount != 100.0:
+                unit_price = price_subtotal / ((1 - (line.discount or 0.0) / 100.0) * abs(line.quantity))
+            else:
+                if self.currency_id.name == 'EUR' or not reverse_charge:
+                    unit_price = line.price_unit
+                else:
+                    eur = self.env.ref('base.EUR')
+                    unit_price = self.currency_id._convert(from_amount=line.price_unit, to_currency=eur, company=self.company, date=self.invoice_date, round=True)
+
+            description = line.name
+            if not document_type == 'TD02':
+                if line.price_subtotal < 0:
+                    moves = line._get_downpayment_lines().move_id
+                    if moves:
+                        description += ', '.join([move.name for move in moves])
+
+            line_dict = {
+                'line': line,
+                'line_number': num + 1,
+                'description': description or 'NO NAME',
+                'unit_price': unit_price,
+                'subtotal_price': price_subtotal,
+            }
+            invoice_lines.append(line_dict)
+
+        # List of dictionaries passed to the template for the part of 'DatiReipilogo'
+        tax_lines = []
```

Comment:
Also would make this section a function, so that it can be adjusted and better refactored (or as Josse says "tidied up") later.